// @coder/nbin/out/common/footer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
/**
 * Reads a footer from a file descriptor with the size of the file
 */
exports.readFooter = function (fd, size) {
    var buffer = Buffer.allocUnsafe(16);
    fs.readSync(fd, buffer, 0, 16, size - 16);
    var headerLength = buffer.readUInt32BE(0);
    var headerOffset = buffer.readUInt32BE(4);
    var contentLength = buffer.readUInt32BE(8);
    var contentOffset = buffer.readUInt32BE(12);
    return {
        headerLength: headerLength,
        headerOffset: headerOffset,
        contentLength: contentLength,
        contentOffset: contentOffset,
    };
};
/**
 * Creates a footer defining byte lengths and offsets
 */
exports.createFooter = function (headerLength, headerOffset, contentLength, contentOffset) {
    var buffer = Buffer.allocUnsafe(16);
    buffer.writeUInt32BE(headerLength, 0);
    buffer.writeUInt32BE(headerOffset, 4);
    buffer.writeUInt32BE(contentLength, 8);
    buffer.writeUInt32BE(contentOffset, 12);
    return buffer;
};
